### ISSUE TITLE
Use tmpfs for overlay upperdir

### ISSUE THREAD

**Issue by Jongy** [moby/moby#41110]
From my experience, the time IO-intensive work takes (like building large projects) can be improved by running it in `tmpfs` (especially for the writes, like build outputs) instead of a disk-backed filesystem. Others use this trick as well; for example, Chromium notes it [in their build tips](https://chromium.googlesource.com/chromium/src/+/0e94f26e8/docs/linux_faster_builds.md#Using-tmpfs).

I use Docker for almost everything I build, and for a long while I've been adding `tmpfs` mounts manually to my containers, like `mount -t tmpfs none /var/lib/docker/overlay2/.../merged/some/tmp/dir`. Then I found the `--tmpfs` flag to `docker run` which is indeed very useful :)

Now, lately I thought - when I run temporary containers, just for build purposes, and when I run any other volatile container (like the containers used for different layers during `docker build`) - is there any reason not to make their entire root filesystem a `tmpfs`? By this, I mean to use a `tmpfs` as the `upperdir` (and `workdir`) of `overlay`. This way, I won't even have to redirect build outputs to a specific folder. Everything I do on the container is really on `tmpfs`, even stuff like `apt-get install`. Very convenient.

I hacked a quick PoC, you can find it [here](https://github.com/Jongy/moby/tree/overlayfs-upperdir-tmpfs). It forces the `diff` and `work` dirs of executed containers to reside on a `tmpfs`. Works fine on my machine.

So basically this is a feature request, because it would be useful to me. I was thinking we can add a `--root-tmpfs` flag to `docker run`. This flag will also require `--rm` to be used (since the container is volatile). It will then use a `tmpfs` for the `upperdir`.

If you guys decide it's worth it, I'd be happy to work on this towards mainlining (incorporating it into the API, updating Docker's CLI, writing tests etc...)

**Comment by the8472** [comment]
On recent kernels (>= 5.10) using the [volatile mount option](https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html#volatile-mount) on the overlayfs should provide most of the performance benefits without risking an OOM.
Kernel maintainers report a 20% speedup for some of their use-cases. https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=c86243b090bc25f81abe33ddfa9fe833e02c4d64

If the upper is stored on ext4 you may want to disable some durability features there too, such as data ordering or auto_da_alloc.

**Comment by solaris-ventus** [comment]
Can you give an example of how to do this for a specific container? Since I did not find that this option could be passed at the docker config level or a separate container. Мои цели такие же как у @Jongy . I used Gentoo
Thank you in advance

**Comment by solaris-ventus** [comment]
My question was exactly about the option 'volatile'.

**Comment by Jongy** [comment]
Ah, that's better :) Umm, from a quick look in the current version of the code, I see no option of passing `volatile` (or any other "general" options. So I think it's impossible. But better ask one of Docker's devs. I see no other issues about the topic - perhaps it's best to raise a direct feature request to support `volatile`.

**Comment by the8472** [comment]
The volatile mount option has to be set at creation time of the overlayfs. To work around the issue that docker doesn't provide a way to pass custom mount parameters you can create an additional overlayfs stack inside the container during startup.

**Comment by Jongy** [comment]
I tried `mount -o remount,volatile /var/lib/docker/overlay2/.../merged` but it doesn't seem like you can change it in runtime (it doesn't give any error, but doesn't change the output of `mount`). It makes sense that you can't do it tho :sweat_smile: 

Do note that you'll need to give certain privileges for the container in order to be able to create mounts inside it (by default, containers can't create new mounts)

**Comment by jasmas** [comment]
This would be useful on raspberry pi where certain containers can cause excessive writes to the MMC in /var/lib/docker

**Comment by Jongy** [comment]
@matti this will cause ALL containers to be volatile (and not just containers, also images, ...)

**Comment by matti** [comment]
@Jongy yes I know, it is wonderful! no more docker kernel hangs!
